popovers: Refactor resend and revoke invitation confirmation modals.

saubhagya-patel opened the pull request:
<!-- Describe your pull request here.-->
This PR refactors the confirmation modals for resending and revoking invitations.

It includes the following changes:
- Use `dialog_widget` instead of `confirm_dialog` to launch confirmation modals for resending and revoking invitations. 
- Change the error text as stated in #31395.
- Associate each modal with a specific ID.
    - `resend_invite_modal`: for the resend invitation confirmation modal.
    - `revoke_invite_modal`: for the revoke invitation confirmation modal.
- Display errors encountered during resending or revoking invitations within the confirmation modal, keeping the modal open.
- <details>
    <summary>Stop the process of revoking or resending invitations if an error occurs.</summary>
    
    For example, earlier, in the case of revoking an invitation, the invitation was revoked disregarding the error if `modal_invite_id !== invite_id || modal_is_multiuse !== is_multiuse` from the `settings_invites.ts` file.  
    Now, if an error occurs during the process of resending or revoking an invitation, it will be displayed inside the confirmation modal, and the process will be discontinued to prevent further actions.
</details>

  
Fixes: #31395.

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**
 **Before Changes:**

<details>
  <summary>1. Resend Invitation Modal</summary>
  
  ![Resend Before](https://github.com/user-attachments/assets/556de01e-6d84-4f06-8614-e2b8e8e7cf11)
  
</details>

<details>
  <summary>2. Revoke Invitation Modal</summary>

  ![Revoke Invitation Before](https://github.com/user-attachments/assets/7b411d51-2eb3-4b2a-b6fb-1366d2c6edf4)

</details>

<details>
  <summary>3. Revoke Invitation Modal (for Invitation Link)</summary>

  ![Revoke Link Before](https://github.com/user-attachments/assets/266fb4f0-a9f3-453e-a7d9-89899e694be0)

</details>

 **After Changes:**

<details>
  <summary>1. Resend Invitation Modal</summary>

  ![Resend After](https://github.com/user-attachments/assets/7038a338-3fe2-49a9-b55a-8447a2b24084)

</details>

<details>
  <summary>2. Revoke Invitation Modal</summary>

  ![Revoke Invitation After](https://github.com/user-attachments/assets/ff6fd697-ea0a-4722-84da-261c9c5a63fc)

</details>

<details>
  <summary>3. Revoke Invitation Modal (for Invitation Link)</summary>

  ![Revoke Link After](https://github.com/user-attachments/assets/dde9d23d-f645-4847-aa1f-29b586e85bed)

</details>

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

alya commented:
Thanks! Please complete the self-review checklist, add the "maintainer review" label using Zulipbot, and mention @sahil839 for review.

saubhagya-patel commented:
@zulipbot add "maintainer review"

saubhagya-patel commented:
Following feedback from Alya Abbott:
> mention @sahil839 for review.

@sahil839 can you please review this PR. Thank you.

sahil839 commented on web/src/settings_invites.ts:
```diff
@@ -254,11 +258,14 @@ export function on_load_success(
         const html_body = render_settings_revoke_invite_modal(ctx);
 
-        confirm_dialog.launch({
+        dialog_widget.launch({
```
I guess you changed this as `close_on_submit` is passed as `true` and we do not want the modal to close immediately. I think you can just pass `close_on_submit`, and other arguments like `id` as well, when using `confirm_dialog` as well.

saubhagya-patel replied:
Yes, you are right about the reason for using `dialog_widget` instead of `confirm_dialog`.
I followed the approach mentioned earlier, but it caused the modal to remain open. I’ve included a screen recording demonstrating the issue with `confirm_dialog`, where `close_on_submit` is passed as `true` and no errors are encountered. Under normal circumstances, the modal should close, but it does not.
<details>
<summary>Screen Recording</summary>

![revoke test](https://github.com/user-attachments/assets/a4243cfb-c6f0-47d9-8f53-8467983d0774)

</details>

Please note that in the above recording, the invite was revoked, but the modal only closed when I clicked the Cancel button. The same case applies to resending an invitation.

sahil839 replied:
Ok, I get this now. It is not the issue with using `confirm_dialog` or `dialog_widget`. When passing `close_on_submit: false`, you need to manually handle closing of modal, which you are doing.

sahil839 replied:
Can we not use `confirm_dialog` as before?

sahil839 commented on web/src/settings_invites.ts:
```diff
@@ -191,12 +193,14 @@ function do_resend_invite({$row, invite_id}: {$row: JQuery; invite_id: string}):
+        return;
     }
 
+    dialog_widget.close();
```
I don't get why is this needed.

saubhagya-patel replied:
I used `dialog_widget`(instead of `confirm_dialog`), which does not close automatically since `close_on_submit` is not passed as `true`. Therefore, I had to call `dialog_widget.close();` to close the modal when no errors are encountered.

Thank you for the review. Please let me know if any further adjustments are needed.

sahil839 replied:
Ohh. Yes, this seems fine for now. But we should just move this to call to `success` function of the API call, if we decide to show to error returned by server in modal as well.

sahil839 reviewed (commented):
Posted a couple of comments.

Maybe we should also show the errors returned by server in the modal. This can be a separate commit. Currently the error is shown as below -
![image](https://github.com/user-attachments/assets/cf255d0a-46ae-4835-b3bf-cdcc8142e271)

saubhagya-patel commented:
Following the review, I have updated this PR.
- The `dialog_widget.close();` call to close the confirmation modal, is now moved inside the `success` function of the API call.
- Now, the errors returned from the server are also shown inside the confirmation modal for resending and revoking invitations, keeping the modal open (I have kept this as a separate commit).

**Screenshots and screen captures:**
**Before Changes:**
<details>
<summary>Resend Invitation Confirmation Modal (Error: No such invitation.)</summary>

![resend server before](https://github.com/user-attachments/assets/f4160ada-15cc-4b85-a0f5-15460af23c6d)

</details>
<details>
<summary>Revoke Invitation Confirmation Modal (Error: No such invitation.)</summary>

![revoke server before](https://github.com/user-attachments/assets/7a0bca4e-7ffa-4d1a-aad3-aad5ffb12934)

</details>
<details>
<summary>Revoke Invitation Link Confirmation Modal (Error: No such invitation.)</summary>

![revoke link server noSuch before](https://github.com/user-attachments/assets/c1dac374-071d-4b47-a89e-c768acdd8edb)

</details>
<details>
<summary>Revoke Invitation Link Confirmation Modal (Error: Invitation has already been revoked.)</summary>

![revoke link already revoked before](https://github.com/user-attachments/assets/1ddbbb29-e51a-4396-baff-af16b6caec75)

</details>

**After Changes:**
<details>
<summary>Resend Invitation Confirmation Modal (Error: No such invitation.)</summary>

![resend server after](https://github.com/user-attachments/assets/165eaae4-14fd-4d3f-8424-1e393e212fec)

</details>
<details>
<summary>Revoke Invitation Confirmation Modal (Error: No such invitation.)</summary>

![revoke server after](https://github.com/user-attachments/assets/7321e70b-81bf-43e8-b721-cb0485c128f1)

</details>
<details>
<summary>Revoke Invitation Link Confirmation Modal (Error: No such invitation.)</summary>

![revoke link server noSuch after](https://github.com/user-attachments/assets/893e1835-f0d1-4aed-bf36-3f04e732cfa0)

</details>
<details>
<summary>Revoke Invitation Link Confirmation Modal (Error: Invitation has already been revoked.)</summary>

![revoke link already revoked after](https://github.com/user-attachments/assets/3f62b8a9-8beb-475c-8f8e-b63775ed9667)

</details>

saubhagya-patel commented:
Requesting a review from @sahil839.

I would like your input regarding the display of error messages returned from the server during the revocation or resending of invitations under the Actions tab of Invitaions. Currently, we use `ui_report.generic_row_button_error(xhr, $revoke_button);`(for example, in revoke invitation), which removes the Resend and Revoke buttons and displays the error in their place.

Should we continue with this approach, or would you recommend a new method? Please let me know if any changes are needed. Thank you.

sahil839 commented on web/src/settings_invites.ts:
```diff
@@ -176,8 +176,16 @@ function do_revoke_invite({
         url,
         error(xhr) {
             ui_report.generic_row_button_error(xhr, $revoke_button);
```
I guess this should be removed now.

sahil839 commented:
> Requesting a review from @sahil839.
> 
> I would like your input regarding the display of error messages returned from the server during the revocation or resending of invitations under the Actions tab of Invitaions. Currently, we use `ui_report.generic_row_button_error(xhr, $revoke_button);`(for example, in revoke invitation), which removes the Resend and Revoke buttons and displays the error in their place.
> 
> Should we continue with this approach, or would you recommend a new method? Please let me know if any changes are needed. Thank you.

If we are showing the error in the moda, there is no need for showing it by removing the button.

saubhagya-patel commented:
Following the review, I have updated this PR.

As @sahil839 mentioned:
> Can we not use confirm_dialog as before?

If we choose to use `confirm_dialog` (as we did previously), we will need to pass `close_on_submit` to `false` in order to display any encountered errors while keeping the modal open. Therefore, we would need to manually close the modal in the `success` function of the API call.

I propose two possible approaches:

1. The `modals.ts` file provides a function named `close_active_if_any()` that can close any _active_ modal. We can utilize this to close the confirmation modal.
2. The `confirm_dialog.launch()` function returns the unique ID of the launched modal. We can use this ID, pass it to the `do_revoke_invite()` function, and then use the `close_if_open()` function from `modals.ts`, passing in the unique ID to close the modal.

For testing, I have implemented the first approach. Please let me know if any changes are required. Thank you.

sahil839 commented on web/src/settings_invites.ts:
```diff
@@ -175,7 +175,14 @@ function do_revoke_invite({
+                xhr,
+                $("#dialog_error"),
+            );
+            $revoke_button.text($t({defaultMessage: "Error!"}));
```
Don't think there is any need to change the button text now.

sahil839 commented on web/src/settings_invites.ts:
```diff
@@ -185,6 +187,7 @@ function do_revoke_invite({
             $revoke_button.text($t({defaultMessage: "Error!"}));
         },
         success() {
+            modals.close_active_if_any();
```
It is fine to use `dialog_widget.close` here. I just meant to not replace `confirm_dialog.launch` with `dialog_widget.launch`.

sahil839 commented:
The commits are not orderd correctly, since in the first commit the modal will not close. So, the changes to close modal manually should also be in the first change.

saubhagya-patel commented:
> The commits are not split correctly, since in the first commit the modal will not close on success. So, the changes to close modal manually should also be in the first commit.

I apologize for the inconvenience caused. I have updated this PR following the review. @sahil839, please let me know if any further changes are needed. Thank you.

sahil839 commented on web/src/settings_invites.ts:
```diff
@@ -259,6 +274,8 @@ export function on_load_success(
                 : $t_html({defaultMessage: "Revoke invitation to {email}"}, {email}),
             html_body,
+            id: "revoke_invite_modal",
+            close_on_submit: false,
```
Would be good to pass `loading_spinner: true` as well since modal is opened while the request is being processed.

sahil839 replied:
And then we would also not need the "Working" text shown in the resend and revoke button -
![image](https://github.com/user-attachments/assets/bec873ad-261a-45e3-ae5f-b39e723926e3)

saubhagya-patel replied:
Okay, so should I remove the code that sets the "Working..." text to the resend and revoke button?
For example, the following line of code, sets the "Working..." text to revoke button:
```
$revoke_button.prop("disabled", true).text($t({defaultMessage: "Working…"}));
```

saubhagya-patel commented:
Following the review, I have updated this PR.
I made the following changes:
- Passed `loading_spinner: true` when launching the `confirm_dialog` to show a loading spinner.
- The "Working..." button text, which was previously displayed while processing the request to resend or revoke an invitation, is no longer shown.

Requesting a review from @sahil839.
Please let me know if any further changes are needed. Thank you.

sahil839 commented on web/src/settings_invites.ts:
```diff
@@ -197,13 +205,21 @@ function do_resend_invite({$row, invite_id}: {$row: JQuery; invite_id: string}):
         );
     }
 
-    $resend_button.prop("disabled", true).text($t({defaultMessage: "Working…"}));
```
I guess we still want the disabled state of button as we do not allow resending the invite again. But I am not sure. @alya thoughts?

Currently in main, the button is disabled after re-sending the mail with button text being changed to "Sent!".

alya replied:
I'm not sure I understand the full context, but the issue did not specify any changes to button states.

alya replied:
Let me know if I'm missing something where such a change would make sense, though.

sahil839 replied:
The PR also changes the errors returned by server in the modal now which is consistent with the other places using modal. Previously the error was shown on the button, i.e. the resend/revoke button.

And previously the button text was changed to "Working..." and the button was disabled while the request was being processed. That is not needed now since modal is closed only after the request completes and till then loading spinner is shown on the modal button.

So, while making the change to no longer make the button text "Working...", the PR also changes the behavior of button once the "Resend" request is completed successfully. Previously the "Resend" button was disabled (background color of button is set to gray) with the text being changed to "Sent!" on successfully resending the invite. So, just wanted to confirm whether we want to keep that behavior or not.

saubhagya-patel replied:
I have updated the PR, so that the button state(in both Resend and Revoke) remains disabled.
Following the above comment, I would like to state a few points.

When we resend an invitation, the following cases may occur:
- **Error due to non-matching fields:** This error is modified to display in the confirmation modal as stated in the issue. It is completed successfully.
- **Server error:** This error was shown in place of the Resend and Revoke buttons. As @sahil839 [mentioned ](https://github.com/zulip/zulip/pull/31896#pullrequestreview-2354556737) server errors should also be displayed in the confirmation modal, I have made changes so that server errors should also display in the confirmation modal. (Please see [this comment](https://github.com/zulip/zulip/pull/31896#issuecomment-2404899338) for changes in how server errors are handled as per this PR.)
- **Successful response from the server:**
  - In main: the Resend button is disabled after resending the invitation, with the button text being changed to "Sent!".
  - In this PR: I have removed the "Working..." text from the button, as the modal remains open while the request is processed, as per [this comment](https://github.com/zulip/zulip/pull/31896#discussion_r1798900111).

I believe we should maintain the disabled state of the Resend button. I have made this change in the PR and pushed it for testing and review.
Thank you.

sahil839 commented on web/src/settings_invites.ts:
```diff
@@ -160,10 +160,11 @@ function do_revoke_invite({
+                defaultMessage: "Error: Could not revoke invitation.",
             }),
-            $("#home-error"),
+            $("#revoke_invite_modal #dialog_error"),
```
Loading spinner is shown even after showing this error. It needs to be hidden.

saubhagya-patel replied:
I apologize for the inconvenience caused. I will update this in the next push, waiting for a final decision on the disabled state of the Resend button.

sahil839 reviewed (commented):
Posted a couple of comments. Also, I think `settings` would be a better prefix for the commit messages than `popovers`.

sahil839 commented on web/src/settings_invites.ts:
```diff
@@ -164,9 +165,10 @@ function do_revoke_invite({
             }),
             $("#home-error"),
         );
+        dialog_widget.hide_dialog_spinner();
```
This should be in the commit which changes the error to be shown in the modal. Also, I guess would be good to have that as a first commit since that contains the changes expected in the issue.

sahil839 replied:
The first commit should only contain changes for showing the errors in dialog instead of `home-error` element and no other changes.

sahil839 commented on web/src/settings_invites.ts:
```diff
@@ -164,9 +165,10 @@ function do_revoke_invite({
     }
 
-    $revoke_button.prop("disabled", true).text($t({defaultMessage: "Working…"}));
+    $revoke_button.prop("disabled", true);
```
Don't think there is a need to disable this button. We should just disable the button when the invite is sent successfully to retain the original behavior.

sahil839 reviewed (commented):
Posted a couple of comments. Should probably be ready for testing by Alya after these changes.

saubhagya-patel commented:
Following the review, I have updated this PR with the following changes:

- I reordered the commits as indicated in [this comment](https://github.com/zulip/zulip/pull/31896#discussion_r1804060824).
- As indicated [here](https://github.com/zulip/zulip/pull/31896#discussion_r1804060541), the Revoke button is no longer disabled. However, the Resend button is disabled after resending the invite to maintain the original behavior.

Requesting a review from @sahil839.
Please let me know if any further changes are needed. Thank you.

sahil839 commented on web/src/settings_invites.ts:
```diff
@@ -282,6 +298,9 @@ export function on_load_success(
             html_body,
+            id: "resend_invite_modal",
+            close_on_submit: false,
+            loading_spinner: true,
```
It can be fine with adding loading spinner in the next commit since client check should be quick, but this is fine too.

sahil839 commented on web/src/settings_invites.ts:
```diff
@@ -206,13 +209,18 @@ function do_resend_invite({$row, invite_id}: {$row: JQuery; invite_id: string}):
     }
 
-    $resend_button.prop("disabled", true).text($t({defaultMessage: "Working…"}));
+    $resend_button.prop("disabled", true);
```
Think button should be disabled only after the resent succeeded, since otherwise the button is disabled even if server returns an error.

saubhagya-patel commented:
Following the review, I have updated this PR with the following changes:
- Moved `loading-spinner` in second commit as indicated in [this comment](https://github.com/zulip/zulip/pull/31896#discussion_r1806501646).
- As indicated [here](https://github.com/zulip/zulip/pull/31896#discussion_r1806503511), the Resend button is disabled only after the process of resending the invitation succeeds.

Requesting a review from @sahil839.
Please let me know if any further changes are needed. Thank you.

sahil839 commented:
Looks good to me.

@alya can you have a look at once. I will summarize the changes once -
- Changes mentioned in issue #31395 are done in the first commit.
- Second commit does the changes to show the errors returned by server in the modal itself which were previously shown in the button, which was used for resending/revoking the invites.

alya commented:
Sounds reasonable to me, thanks! I'm not sure how to test this efficiently, and would prefer for @timabbott to review next.

alya commented:
@timabbott approved by @sahil839 .

timabbott commented:
Merged, thanks @saubhagya-patel! I hope you learned a lot about creating a readable PR from this review process :)

saubhagya-patel commented:
My pleasure, @timabbott. I admit that I have learned a great deal from the review process. I would also like to extend my gratitude to @sahil839 and @alya for their valuable feedback. I'm looking forward to contributing more in the future.
